File: labs/07-agent-evaluation/go/main.go
package main

import (
	"encoding/json"
	"fmt"
	"strings"

	openaiclient "github.com/ZYZ-Labs/agent-labs/shared/config"
)

type testCase struct {
	Input             string   `json:"input"`
	ExpectedKeywords  []string `json:"expected_keywords"`
	Reference         string   `json:"reference"`
}

var testCases = []testCase{
	{
		Input:            "What port range is safe for user services?",
		ExpectedKeywords: []string{"1024", "65535"},
		Reference:        "User services should use ports from 1024 to 65535.",
	},
	{
		Input:            "Explain idempotency in one sentence.",
		ExpectedKeywords: []string{"same", "multiple", "result"},
		Reference:        "Idempotency means calling an operation multiple times produces the same result.",
	},
}

func ruleCheck(answer string, keywords []string) map[string]interface{} {
	var missing []string
	lower := strings.ToLower(answer)
	for _, kw := range keywords {
		if !strings.Contains(lower, strings.ToLower(kw)) {
			missing = append(missing, kw)
		}
	}
	return map[string]interface{}{"passed": len(missing) == 0, "missing": missing}
}

func llmJudge(client *openaiclient.Client, answer, reference string) map[string]interface{} {
	prompt := fmt.Sprintf(`Rate how well the following answer matches the reference answer.
Answer: %s
Reference: %s
Respond with JSON only: {"score": 1-10, "reason": "..."}`, answer, reference)
	temp := 0.0
	resp, err := client.ChatCompletion(openaiclient.ChatCompletionRequest{
		Messages: []openaiclient.Message{{Role: "user", Content: prompt}},
		ResponseFormat: map[string]string{"type": "json_object"},
		MaxTokens: intPtr(200),
		Temperature: &temp,
	})
	if err != nil {
		return map[string]interface{}{"score": 0, "reason": fmt.Sprintf("judge call failed: %v", err)}
	}
	var result map[string]interface{}
	_ = json.Unmarshal([]byte(resp.Choices[0].Message.Content), &result)
	return result
}

func evaluateAgent(client *openaiclient.Client) []map[string]interface{} {
	var results []map[string]interface{}
	for _, c := range testCases {
		answer := simpleAgent(client, c.Input)
		ruleResult := ruleCheck(answer, c.ExpectedKeywords)
		judgeResult := llmJudge(client, answer, c.Reference)
		results = append(results, map[string]interface{}{
			"input":            c.Input,
			"answer":           answer,
			"rule_passed":      ruleResult["passed"],
			"missing_keywords": ruleResult["missing"],
			"judge_score":      judgeResult["score"],
			"judge_reason":     judgeResult["reason"],
		})
	}
	return results
}

func simpleAgent(client *openaiclient.Client, question string) string {
	temp := 0.0
	resp, err := client.ChatCompletion(openaiclient.ChatCompletionRequest{
		Messages: []openaiclient.Message{{Role: "user", Content: question}},
		MaxTokens: intPtr(200),
		Temperature: &temp,
	})
	if err != nil {
		return fmt.Sprintf("Error: %v", err)
	}
	return resp.Choices[0].Message.Content
}

func intPtr(i int) *int {
	return &i
}

func main() {
	client := openaiclient.NewClient()
	results := evaluateAgent(client)

	passed := 0
	total := len(results)
	var scoreSum float64
	for _, r := range results {
		if r["rule_passed"].(bool) {
			passed++
		}
		if s, ok := r["judge_score"].(float64); ok {
			scoreSum += s
		}
	}
	avg := scoreSum / float64(total)

	b, _ := json.MarshalIndent(results, "", "  ")
	fmt.Println(string(b))
	fmt.Printf("\nRule checks passed: %d/%d\n", passed, total)
	fmt.Printf("Average judge score: %.1f/10\n", avg)
}

File: labs/07-agent-evaluation/java/Main.java
import com.agentlabs.OpenAiClient;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class Main {

    private static final List<Map<String, Object>> TEST_CASES = List.of(
            Map.of(
                    "input", "What port range is safe for user services?",
                    "expected_keywords", List.of("1024", "65535"),
                    "reference", "User services should use ports from 1024 to 65535."
            ),
            Map.of(
                    "input", "Explain idempotency in one sentence.",
                    "expected_keywords", List.of("same", "multiple", "result"),
                    "reference", "Idempotency means calling an operation multiple times produces the same result."
            )
    );

    public static void main(String[] args) throws Exception {
        OpenAiClient client = new OpenAiClient();
        ObjectMapper mapper = new ObjectMapper();

        List<Map<String, Object>> results = evaluateAgent(client, mapper);

        long passed = results.stream().filter(r -> (Boolean) r.get("rule_passed")).count();
        int total = results.size();
        double avgScore = results.stream()
                .mapToDouble(r -> r.get("judge_score") instanceof Number n ? n.doubleValue() : 0.0)
                .average().orElse(0.0);

        System.out.println(mapper.writerWithDefaultPrettyPrinter().writeValueAsString(results));
        System.out.println("\nRule checks passed: " + passed + "/" + total);
        System.out.printf("Average judge score: %.1f/10%n", avgScore);
    }

    private static List<Map<String, Object>> evaluateAgent(OpenAiClient client, ObjectMapper mapper) throws Exception {
        List<Map<String, Object>> results = new ArrayList<>();
        for (Map<String, Object> testCase : TEST_CASES) {
            String answer = simpleAgent(client, mapper, (String) testCase.get("input"));
            Map<String, Object> ruleResult = ruleCheck(answer, (List<String>) testCase.get("expected_keywords"));
            Map<String, Object> judgeResult = llmJudge(client, mapper, answer, (String) testCase.get("reference"));

            results.add(Map.of(
                    "input", testCase.get("input"),
                    "answer", answer,
                    "rule_passed", ruleResult.get("passed"),
                    "missing_keywords", ruleResult.get("missing"),
                    "judge_score", judgeResult.getOrDefault("score", null),
                    "judge_reason", judgeResult.getOrDefault("reason", "")
            ));
        }
        return results;
    }

    private static Map<String, Object> ruleCheck(String answer, List<String> keywords) {
        List<String> missing = new ArrayList<>();
        String lower = answer.toLowerCase();
        for (String kw : keywords) {
            if (!lower.contains(kw.toLowerCase())) {
                missing.add(kw);
            }
        }
        return Map.of("passed", missing.isEmpty(), "missing", missing);
    }

    private static Map<String, Object> llmJudge(OpenAiClient client, ObjectMapper mapper, String answer, String reference) throws Exception {
        String prompt = "Rate how well the following answer matches the reference answer.\nAnswer: " + answer
                + "\nReference: " + reference
                + "\nRespond with JSON only: {\"score\": 1-10, \"reason\": \"...\"}";

        ObjectNode payload = mapper.createObjectNode();
        ArrayNode messages = mapper.createArrayNode();
        ObjectNode user = messages.addObject();
        user.put("role", "user");
        user.put("content", prompt);
        payload.set("messages", messages);
        ObjectNode responseFormat = mapper.createObjectNode();
        responseFormat.put("type", "json_object");
        payload.set("response_format", responseFormat);
        payload.put("max_tokens", 200);
        payload.put("temperature", 0.0);

        JsonNode response = client.chatCompletion(payload);
        String content = client.extractMessage(response).get("content").asText();
        return mapper.readValue(content, Map.class);
    }

    private static String simpleAgent(OpenAiClient client, ObjectMapper mapper, String question) throws Exception {
        ObjectNode payload = mapper.createObjectNode();
        ArrayNode messages = mapper.createArrayNode();
        ObjectNode user = messages.addObject();
        user.put("role", "user");
        user.put("content", question);
        payload.set("messages", messages);
        payload.put("max_tokens", 200);
        payload.put("temperature", 0.0);

        JsonNode response = client.chatCompletion(payload);
        return client.extractMessage(response).get("content").asText();
    }
}
